app/src/main/kotlin/com/flxrs/dankchat/preferences/userdisplay/UserDisplayDto.kt

```diff
@@ -0,0 +1,97 @@
+@file:OptIn(ExperimentalContracts::class)
+
+package com.flxrs.dankchat.preferences.userdisplay
+
+import androidx.annotation.ColorInt
+import com.flxrs.dankchat.data.database.UserDisplayEntity
+import kotlinx.serialization.Serializable
+import kotlin.contracts.ExperimentalContracts
+import kotlin.contracts.contract
+
+@Serializable
+data class UserDisplayDto(
+    val id: Int,
+    val username: String,
+    val enabled: Boolean,
+    val colorEnabled: Boolean,
+    val color: Int,
+    val aliasEnabled: Boolean,
+    val alias: String,
+) {
+    companion object {
+        fun UserDisplayItem.Entry.toDto() = UserDisplayDto(
+            id = id,
+            // trim input entries (especially important for username matching)
+            username = username.trim(),
+            enabled = enabled,
+            colorEnabled = colorEnabled,
+            color = color,
+            aliasEnabled = aliasEnabled,
+            alias = alias.trim(),
+        )
+
+        fun UserDisplayDto.toEntryItem() = UserDisplayItem.Entry(
+            id = id,
+            username = username,
+            enabled = enabled,
+            colorEnabled = colorEnabled,
+            color = color,
+            aliasEnabled = aliasEnabled,
+            alias = alias,
+        )
+
+        fun UserDisplayEntity.toDto() = UserDisplayDto(
+            id = id,
+            username = targetUser,
+            enabled = enabled,
+            colorEnabled = colorEnabled,
+            color = color,
+            aliasEnabled = aliasEnabled,
+            alias = alias.orEmpty()
+        )
+    }
+
+    fun toEntity() = UserDisplayEntity(
+        id = id,
+        targetUser = username,
+        enabled = enabled,
+        colorEnabled = colorEnabled,
+        color = color,
+        aliasEnabled = aliasEnabled,
+        alias = alias
+    )
+}
+
+// these are defined as extension type to allow convenient checking even the object is null
+/** return whether alias is set, and enabled
+ * calling on null will return false */
+fun UserDisplayDto?.hasAlias(): Boolean {
```

Comment:
I would like to see some of this validation logic handled by the repository, if possible, similar to how I only add valid highlights/ignores to messages. The ui should just get a static "state" which already contains everything needed.

Reply:
I have gone with the way mentioned above (new data class `UserDisplayEffectiveValue`). There will similar methods on the class `UserDisplayEffectiveValue` tho, but the purpose is just to simplify null checking / handling (similar purpose like String's `ifEmpty`).

I also move the logic of formatting username / color to `Message` classes.